server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -1051,14 +1051,17 @@ public Symbol visitObjectLiteral(ObjectLiteral node, ExpressionAnalysisContext c
                 return Literal.EMPTY_OBJECT;
             }
             List<Symbol> arguments = new ArrayList<>(values.size() * 2);
+            var typeBuilder = ObjectType.of(ColumnPolicy.DYNAMIC);
             for (Map.Entry<String, Expression> entry : values.entrySet()) {
                 arguments.add(Literal.of(entry.getKey()));
                 arguments.add(entry.getValue().accept(this, context));
+                typeBuilder.setInnerType(entry.getKey(), entry.getValue().accept(this, context).valueType());
             }
-            return allocateFunction(
-                MapFunction.NAME,
+            return new Function(
+                MapFunction.SIGNATURE,
                 arguments,
-                context);
```

Comment:
There could be a better way; while resolving arg types/return type for MapFunction, inner types of the returning object type can be deduced?